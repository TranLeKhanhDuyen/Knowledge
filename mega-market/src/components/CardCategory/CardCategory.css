.card-category-container {
  display: flex;
  flex-direction: column;
  align-items: center;
  cursor: pointer;
}

.card-category-container-circle {
  max-width: 172px;
  height: auto;
}

.category-img-wrapper {
  display: flex;
  align-items: center;
  justify-content: center;
  border: 1px solid transparent;
  background-color: var(--secondary-color);
  width: 132px;
  height: 132px;
  max-width: 100%;
  max-height: 100%;
}

.category-img {
  width: 100%;
  height: 100%;
  object-fit: contain;
}

.category-img-wrapper:hover {
  border-color: var(--primary-color);
  box-shadow: 0 4px 8px rgba(0, 0, 0, 0.2);
}

.category-img-wrapper-circle {
  border-radius: 50%;
  padding: 17px;
}

.category-img-wrapper-square {
  border-radius: 16px;
  padding: 20px;
}

.category-details {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 4px;
  padding: 12px;
  width: 100%;
}

.category-name-square {
  font-size: var(--font-md);
  color: var(--text-secondary-color);
  font-weight: var(--fw-bold);
}

.category-discount-square {
  font-size: var(--font-lg);
  color: var(--text-primary-color);
  font-weight: var(--fw-bold);
}
