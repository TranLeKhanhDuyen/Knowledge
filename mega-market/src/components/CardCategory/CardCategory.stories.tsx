import type { Meta, StoryObj } from '@storybook/react'

import { CardCategory } from '@components'
import img1 from '@assets/images/products/phone-2.svg'
import img2 from '@assets/images/products/cherry.svg'

export default {
  title: 'Components/CardCategory',
  component: CardCategory,
  tags: ['autodocs'],
  argTypes: {
    additionalClass: { description: 'Add class to the category' },
    imageUrl: { description: 'imageUrl content of category' },
    alt: { description: 'alt content of category' },
    name: { description: 'name content of category' },
    discountPercent: { description: 'discount percent content of category' }
  }
} as Meta

type Story = StoryObj<typeof CardCategory>

export const Circle: Story = {
  args: {
    imageUrl: img1,
    alt: 'phone',
    name: 'Galaxy S22 Ultra',
    variant: 'circle'
  }
}

export const Square: Story = {
  args: {
    imageUrl: img2,
    variant: 'square',
    name: 'Daily Essentials',
    discountPercent: '50%'
  }
}
