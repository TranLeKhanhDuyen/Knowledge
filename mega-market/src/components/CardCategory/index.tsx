import React from 'react'
import './CardCategory.css'

type TCardCategoryVariant = 'square' | 'circle'

interface ICardCategoryProps {
  imageUrl: string
  alt?: string
  name?: string
  discountPercent?: string
  additionalClass?: string
  variant?: TCardCategoryVariant
}

const CardCategory: React.FC<ICardCategoryProps> = ({
  imageUrl,
  alt,
  name,
  discountPercent,
  variant = 'circle',
  additionalClass
}) => {
  return (
    <article
      className={`card-category-container card-category-container-${variant} ${additionalClass}`}
    >
      <figure
        className={`category-img-wrapper category-img-wrapper-${variant} ${additionalClass}`}
      >
        <img className='category-img' src={imageUrl} alt={`Image at ${alt}`} />
      </figure>
      <aside className={`category-details ${additionalClass}`}>
        <p className={`category-name category-name-${variant}`}>{name}</p>
        {discountPercent && (
          <p
            className={`category-discount category-discount-${variant}`}
          >{`UP to ${discountPercent} OFF`}</p>
        )}
      </aside>
    </article>
  )
}

export default CardCategory
