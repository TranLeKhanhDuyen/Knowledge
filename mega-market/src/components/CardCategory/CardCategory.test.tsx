import { fireEvent, render } from '@testing-library/react'
import CardCategory from '.'

describe('CardCategory component', () => {
  test('renders correctly', () => {
    const { container } = render(
      <CardCategory imageUrl='url' alt='test alt' name='test name' />
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  test('renders name correctly', () => {
    const testName = 'Test Name'
    const { getByText } = render(
      <CardCategory imageUrl='url' alt='test alt' name={testName} />
    )
    expect(getByText(testName)).toBeInTheDocument()
  })

  test('renders discount correctly', () => {
    const testDiscountPercent = '50'
    const { getByText } = render(
      <CardCategory
        imageUrl='url'
        alt='test alt'
        name='test name'
        discountPercent={testDiscountPercent}
      />
    )
    expect(getByText(`UP to ${testDiscountPercent} OFF`)).toBeInTheDocument()
  })

  test('applies additionalClass correctly', () => {
    const additionalClass = 'additional-class'
    const { container } = render(
      <CardCategory
        imageUrl='url'
        alt='test alt'
        name='test name'
        additionalClass={additionalClass}
      />
    )
    expect(container.firstChild).toHaveClass(additionalClass)
  })

  test('calls onClick when image is clicked', () => {
    const onClick = jest.fn()
    const { getByAltText } = render(
      <CardCategory
        imageUrl='url'
        alt='test alt'
        name='test name'
        onClick={onClick}
      />
    )
    fireEvent.click(getByAltText('test alt'))
    expect(onClick).toHaveBeenCalled()
  })
})
