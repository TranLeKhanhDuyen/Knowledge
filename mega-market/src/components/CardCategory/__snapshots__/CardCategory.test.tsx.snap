// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardCategory component renders correctly 1`] = `
<article
  class="card-category-container card-category-container-circle "
>
  <figure
    class="category-img-wrapper category-img-wrapper-circle"
  >
    <img
      alt="test alt"
      class="category-img"
      src="url"
    />
  </figure>
  <aside
    class="category-details"
  >
    <p
      class="category-name category-name-circle"
    >
      test name
    </p>
  </aside>
</article>
`;
